grid.v3/codec/registry.go

```diff
@@ -1,61 +1,90 @@
 package codec
 
 import (
+	"errors"
 	"reflect"
 	"sync"
+
+	"github.com/gogo/protobuf/proto"
 )
 
-var registry *codecRegistry = &codecRegistry{reg: map[string]Codec{}} //shared registry
+var (
+	ErrNonProtoMessage         = errors.New("codec: non proto message")
+	ErrUnregisteredMessageType = errors.New("codec: unregistered message type")
+)
 
-type CodecRegistry interface {
-	Register(v interface{}, codec Codec)
-	GetCodec(v interface{}) (Codec, bool)
-	GetCodecName(name string) (Codec, bool)
-}
+var (
+	mu       = &sync.RWMutex{}
+	registry = map[string]interface{}{}
+)
 
-func Registry() CodecRegistry {
-	return registry
-}
+func Register(v interface{}) error {
+	mu.Lock()
+	defer mu.Unlock()
 
-type codecRegistry struct {
-	reg   map[string]Codec
-	mutex sync.RWMutex
-}
+	// The value 'v' must not be registered
+	// as a pointer type, but to check if
+	// it is a proto message, the pointer
+	// type must be checked.
+	pv := reflect.New(reflect.TypeOf(v)).Interface()
+
+	_, ok := pv.(proto.Message)
+	if !ok {
+		return ErrNonProtoMessage
+	}
 
-func (cr *codecRegistry) Register(v interface{}, codec Codec) {
-	n := Name(v)
-	//fmt.Println("reg:", n)
-	cr.mutex.Lock()
-	defer cr.mutex.Unlock()
-	cr.reg[n] = codec
+	name := TypeName(v)
+	registry[name] = v
+	return nil
 }
 
-func (cr *codecRegistry) GetCodec(v interface{}) (Codec, bool) {
-	return cr.GetCodecName(Name(v))
+func Marshal(v interface{}) (string, []byte, error) {
+	mu.RLock()
+	defer mu.RUnlock()
+
+	name := TypeName(v)
+	_, ok := registry[name]
+	if !ok {
+		return "", nil, ErrUnregisteredMessageType
+	}
+	buf, err := protoMarshal(v)
+	if err != nil {
+		return "", nil, err
+	}
+	return name, buf, nil
 }
 
-func (cr *codecRegistry) GetCodecName(n string) (Codec, bool) {
-	//fmt.Println("get:", n)
-	cr.mutex.RLock()
-	defer cr.mutex.RUnlock()
-	c, ok := cr.reg[n]
-	return c, ok
+func Unmarshal(buf []byte, name string) (interface{}, error) {
+	mu.RLock()
+	defer mu.RUnlock()
+
+	c, ok := registry[name]
+	if !ok {
+		return nil, ErrUnregisteredMessageType
+	}
+	v := reflect.New(reflect.TypeOf(c)).Interface()
+	err := protoUnmarshal(buf, v)
+	if err != nil {
+		return nil, err
+	}
+	return v, nil
 }
 
-func Name(v interface{}) string {
+func TypeName(v interface{}) string {
 	rt := reflect.TypeOf(v)
 	name := rt.String()
-	// But for named types (or pointers to them), qualify with import path (but see inner comment).
-	// Dereference one pointer looking for a named type.
-	star := ""
-	if rt.Name() == "" {
-		if pt := rt; pt.Kind() == reflect.Ptr {
-			star = "*"
-			rt = pt
-		}
-	}
-	if rt.Name() != "" {
-		name = star + rt.PkgPath() + "." + rt.Name()
+	if name[0] == '*' {
```

Comment:
I don't think this will work if the types are of the same name but from different packages?